flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/runtime/metadata/ModifiableSavepointMetadata.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.state.api.runtime.metadata;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.checkpoint.MasterState;
+import org.apache.flink.runtime.checkpoint.OperatorState;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.state.api.BootstrapTransformation;
+import org.apache.flink.state.api.runtime.OperatorIDGenerator;
+import org.apache.flink.types.Either;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Savepoint metadata that can be modified.
+ */
+@Internal
+public class ModifiableSavepointMetadata extends SavepointMetadata {
+
+	private transient Map<OperatorID, Either<OperatorState, BootstrapTransformation<?>>> operatorStateIndex;
```

Comment:
Just for the sake of readability, I would introduce a dedicated wrapper class, say `OperatorStateSpec` to replace this `Either`. `Either.left(...)` / `Either.isLeft()` is harder to understand at first glance than `OperatorStateSpec.existing(...) / OperatorStateSpec.isExistingState()`.

Reply:
Concerning `transient`:

I think generally speaking, the `ModifiableSavepointMetadata` should not need to be serializable at all, since its main concern is bookkeeping what existing / new operators there are to facilitate building the job graph, and therefore do not need to be part of the serialized job graph.

I think the only reason it needs to be serializable at the moment is because some runtime classes, e.g. `MergeOperatorStates` reduce operator, holds a reference to it.
It should be possible (as IMO also more reasonable) to reduce the dependencies on `SavepointMetadata` in these runtime classes, and only pass in the necessary information. This should allow us to not make `SavepointMetadata` a `Serializable`.